Release GIL before destructing RPCAgent subclasses.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57030 Assert that GIL is not held in blocking destructors
* **#57029 Release GIL before destructing RPCAgent subclasses.**

Partially addresses https://github.com/pytorch/pytorch/issues/56297

This fixes deadlocks when the threads the RPCAgent are blocking
on try to take the GIL.  This also adds a general utility for
making shared_ptr run destructors without GIL.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28030294](https://our.internmc.facebook.com/intern/diff/D28030294)

ezyang committed:
Release GIL before destructing RPCAgent subclasses.

Partially addresses https://github.com/pytorch/pytorch/issues/56297

This fixes deadlocks when the threads the RPCAgent are blocking
on try to take the GIL.  This also adds a general utility for
making shared_ptr run destructors without GIL.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Release GIL before destructing RPCAgent subclasses."


Partially addresses https://github.com/pytorch/pytorch/issues/56297

This fixes deadlocks when the threads the RPCAgent are blocking
on try to take the GIL.  This also adds a general utility for
making shared_ptr run destructors without GIL.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28030294](https://our.internmc.facebook.com/intern/diff/D28030294)

[ghstack-poisoned]

albanD commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -547,13 +547,14 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
                        int numSendRecvThreads,
                        std::chrono::milliseconds rpcTimeout) {
-        return std::make_unique<ProcessGroupAgent>(
+        return std::shared_ptr<ProcessGroupAgent>(new ProcessGroupAgent(
```
Why not keep a unique_ptr here with a custom deleter?

ezyang replied:
If you look at the class definition above, the class is actually a `shared_ptr`. So what they were previously doing here was creating a unique ptr, and then using the stdlib functionality that lets you steal into a shared pointer. So you could do a unique ptr with a custom deleter but it would be pointless, it's equivalent to what's written here (but what is written here is a lot more direct).

albanD commented on torch/csrc/utils/pybind.h:
```diff
@@ -107,3 +107,52 @@ template<> struct type_caster<at::IntArrayRef> {
+// This function is a valid shared_ptr destructor and can be used to
+// conveniently allocate a shared_ptr to an object whose destructor will be run
+// without the GIL.  Often, this logic cannot be conveniently placed on the
+// destructor itself because the object has no dependency on Python.
```
nit: it wasn't clear to me at the beginning that this logic is for objects that are used either in a pure c++ context or a python-bound context. And so this custom destructor allows to bind the GIL-less destruction to specific instances of the object (namely the python-bound ones).

But I'm not sure how to express that idea in a small sentence :D

ezyang replied:
I didn't have time to write a short letter so I wrote a long letter

albanD reviewed (commented):
Looks good, just a small question.

ezyang committed:
Update on "Release GIL before destructing RPCAgent subclasses."


Partially addresses https://github.com/pytorch/pytorch/issues/56297

This fixes deadlocks when the threads the RPCAgent are blocking
on try to take the GIL.  This also adds a general utility for
making shared_ptr run destructors without GIL.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28030294](https://our.internmc.facebook.com/intern/diff/D28030294)

[ghstack-poisoned]

albanD reviewed (approved):
LGTM!
